clearwater-queue-manager.root/usr/share/clearwater/clearwater-queue-manager/scripts/upgrade_cassandra.py

```diff
@@ -0,0 +1,204 @@
+
+    _log.info("Updating cassandra repo")
+
+    # Create a backup of the sources list file to roll back to if necessary
+    shutil.copyfile(SOURCES_FILE, TEMP_SOURCES_FILE)
+
+    # The clearwater sources list should include a single source. Update it
+    # to point to the new repo directory
+    new_repo = ""
+    with open(SOURCES_FILE, "r") as file:
+        for line in file:
+            if line.startswith("deb"):
+                split_line = line.split(" ")
+                if len(split_line) > 2:
+                    old_repo = split_line[1]
+
+                    # If old_repo has a trailing slash, remove it
+                    if old_repo.endswith("/"):
+                        old_repo = old_repo[:-1]
+
+                    # Check that the Clearwater version isn't changing
+                    split_repo = old_repo.split("/")
+
+                    old_cc_version = split_repo[-1].split("cass")[0]
+                    new_cc_version = repo_dir.split("cass")[0]
+
+                    if old_cc_version != new_cc_version:
+                        _log.error("Can't upgrade CC and Cassandra at the same time")
+                    else:
+                        # Rebuild the string, replacing the final path segment with
+                        # the new repo directory
+                        path_list = split_repo[:-1]
+                        path_list.append(repo_dir)
+                        new_repo = "/".join(path_list)
+                break
+
+    if new_repo == "":
+        _log.error("Aborting upgrade")
+        return False
+
+    # Add the new Cassandra repo to the sources, then write back
+    new_sources = "deb {0} binary/\n".format(new_repo)
+    safely_write(SOURCES_FILE, new_sources)
+
+    # Now, check that we can find the repo and that upgrade will work by doing a dry-run
+    if run_command("sudo apt-get update " +
+                   "-o Dir::Etc::sourcelist=\"sources.list.d/clearwater.list\" " +
+                   "-o Dir::Etc::sourceparts=\"-\" " +
+                   "-o APT::Get::List-Cleanup=\"0\""):
+        _log.error("Unable to contact repo {0}".format(new_repo))
+        shutil.move(TEMP_SOURCES_FILE, SOURCES_FILE)
+        return False
+
+    if run_command("sudo apt-get install --only-upgrade --dry-run " + PACKAGES + " "
+                   "-o Dir::Etc::sourcelist=\"sources.list.d/clearwater.list\" " +
+                   "-o Dir::Etc::sourceparts=\"-\" " +
+                   "-o APT::Get::List-Cleanup=\"0\""):
+        _log.error("Will not be able to upgrade Cassandra packages")
+        shutil.move(TEMP_SOURCES_FILE, SOURCES_FILE)
+        return False
+
+    # Start performing the uprade
+    _log.info("Draining node")
+    if run_command("/usr/share/clearwater/bin/run-in-signaling-namespace nodetool drain"):
+        _log.error("Unable to drain node")
+        shutil.move(TEMP_SOURCES_FILE, SOURCES_FILE)
+        return False
+
+    if run_command("sudo monit stop cassandra_process"):
```

Comment:
`-g cassandra` rather than cassandra_process (this stops the cassandra monitoring scripts too)